### PR TITLE
Remove support for GHC 7.10

### LINKED ISSUE #3719: Remove support for GHC 7.10 (aka base <4.9)
Agda currently [maintains support for GHC 7.10](https://github.com/agda/agda/blob/47eecb71752536effba9f26b337a66f7b2c722a9/Agda.cabal#L36), which was released ~3.5 years ago. That version [corresponds with `base-4.8.*`](https://wiki.haskell.org/Base_package). It is the pre-GHC-8 version still supported.

A lot has changed in Haskell-world in that time, and continues to change, which makes supporting a range of versions difficult and limits what tooling can be used within the Agda codebase.

For example, new features of the GHC runtime and base library could facilitate removing the C preprocessor from most of the source files, which should improve compilation times, and will simplify use of source analysis, refactoring, and formatting tools. See #3718 and #1875.

What is the GHC version support policy for Agda?

Similar older issues for previously-dropped versions:
  * https://github.com/agda/agda/issues/2692 (7.8.4)
  * https://github.com/agda/agda/issues/1577 (7.4.2)
  * https://github.com/agda/agda/issues/1392 (7.0.4)
  * https://github.com/agda/agda/issues/1393 (7.0.x)

### PR DESCRIPTION
**Pull request by rwe**
~**NOTE**: @asr [notes that this should not be included](https://github.com/agda/agda/issues/3719#issuecomment-487043450) in the 2.6.1 patch release. Let me know if this PR should be pointed to a branch other than `master`.~

Fixes #3719 (_Remove support for GHC 7.10 (aka base <4.9_)
Based on/includes #3726 (_Simplify and consolidate GHC version conditions_).
(Please see that first to avoid redundant comments).

The minimum base packages are bumped (only) up to the versions included in the minimum supported GHC compiler, 8.0.1.

It's possible that these should be increased further to 8.0.2 if 8.0.1 isn't _really_ intended to be supported; or that their minimum bounds on those packages should be removed altogether. Discussed further in the relevant commit.

### CODE REVIEW

**Comment by jespercockx** [comment]
@asr said that we should not drop support before 2.6.0.1, but I guess it's fine to drop it for 2.6.1. So it's fine to keep this PR against master (but it will probably not be merged until after the release of 2.6.0.1).

File: Agda.cabal [hunk 1]
```diff
@@ -158,24 +156,23 @@ library
...
+                , bytestring >= 0.10.8.1 && < 0.11
+                , containers >= 0.5.7.1 && < 0.7
                 , data-hash >= 0.2.0.0 && < 0.3
-                , deepseq >= 1.4.1.1 && < 1.5
-                , directory >= 1.2.2.0 && < 1.4
-                -- EdisonCore 1.3.2 doesn't compile with GHC 7.10.*.
-                , EdisonCore == 1.3.1.1 || >= 1.3.2.1 && < 1.4
+                , deepseq >= 1.4.2.0 && < 1.5
+                , directory >= 1.2.6.2 && < 1.4
+                , EdisonCore == >= 1.3.1.1 && < 1.4
```
Comment (UlfNorell) [inline Agda.cabal]: Here's the parse error Travis is complaining about. I guess you want
```suggestion
                , EdisonCore >= 1.3.1.1 && < 1.4
```

**Review by UlfNorell — CHANGES_REQUESTED** [review]

File: Agda.cabal (hunk 1 above)
Comment (rwe) [reply Agda.cabal]: Ah, thank you!

**Comment by rwe** [comment]
@UlfNorell — Unfortunately GitHub isn't giving me an option to re-request review, but whenever you're able, would you please take a look? I think all issues are now addressed.

**Review by UlfNorell — APPROVED** [review]

**Comment by rwe** [comment]
The post-review force push was a commit message rewrite due to the discussion here: https://github.com/agda/agda/pull/3726#issuecomment-487963875

The PR contents themselves are unchanged: https://github.com/erydo/agda/compare/725ec67..e8270b4

**Comment by asr** [comment]
Since we must keep support for GHC 7.10.3 for Agda 2.6.0.1, we need to decide how to proceed for merging this PR. Some options:

A) To merge into the `future` branch, so we can release 2.6.0.1 from `master`.

B) To merge into the `master` branch and

B1) to create a new branch for 2.6.0.1 and cherry-pick into the new branch every 2.6.0.1-commit.

B2) to forget about 2.6.0.1 and release 2.6.1 from `master` very soon (including the bug-fixes for 2.6.0.1 and the non-support for GHC 7.10.3)
 
I think the easy option for every one (including me) is B2.

@erydo, please fix the conflicts reported by GitHub.

**Comment by jespercockx** [comment]
I'm planning to (hopefully) very soon fix the remaining issues with my PR #3589 and merge it into `master`. However, I feel like this could use some more testing to sort out potential performance issues, which could be problematic if we go for option B2. So I would personally prefer option B1 (we can postpone creating the branch until my PR is ready to merge, however).

**Review by asr — CHANGES_REQUESTED** [review]
Please add to the `CHANGELOG.md` entry for 2.6.1:

Installation and infrastructure
------------------------------------------

* Removed support for GHC 7.10.3

**Comment by rwe** [comment]
I'd recommend merging the PRs rather than rebase-applying them like was done with #3726. That allows you to merge the same feature branch into multiple downstream branches (e.g. `master`/`experimental`/`2.6.0-maintenance`) without duplicating +and requiring manual resolution of) the commits. Assuming that this branch was based on a tip that was a common base of those (or at least a future common ancestry).

In any case, this PR has been rebased on `master` for now.

**Comment by rwe** [comment]
@asr Done

**Comment by UlfNorell** [comment]
It also feels wrong to release 2.6.0.1 as 2.6.1 when it really is just a minor bugfix release. There shouldn't be much to cherry-pick, so I'm in favour of B1.

**Review by asr — APPROVED** [review]

**Comment by asr** [comment]
Thanks @erydo!

**Pull request merged**